### ISSUE TITLE
Issues with Self-Play

### ISSUE THREAD

**Issue by niskander** [Unity-Technologies/ml-agents#3563]
**Describe the bug**
Hello, I just started using Self-Play, and while I can't say yet whether it's working or not, I'm seeing a couple of issues which may be affecting learning or may possibly just be logging errors.

1) So even though the teams are set correctly for the agents, and initially the training correctly connects to the two team brains, afterwards only one of the teams appears in the logs. (I've double-checked the teams at runtime and made sure they don't all somehow revert to one team.) You can see that in the screenshot below:
![selfplayteams_LI](https://user-images.githubusercontent.com/1988272/75930652-0e0f7280-5e41-11ea-9f44-9d19a810cabc.jpg)

2) The other issue is with the rewards graphs. I'm only using extrinsic rewards, and the rewards are zero-sum. Before upgrading the plugin and using self-play, the two reward graphs (cumulative reward and extrinsic reward) were identical, but now somehow they're not. (See below; you can also see in the previous screenshot that the mean rewards were all 0s)
![selfplay_test_graph_LI](https://user-images.githubusercontent.com/1988272/75930654-0e0f7280-5e41-11ea-8a43-413d77a512b1.jpg)

3) The third issue and perhaps the most serious is that the training seems to freeze way more often than before. I have to 'Ctrl+C' in the command prompt every once in a while to revive it. This is happening even after eliminating concurrent environments and removing the increased time-scale (so it's supposed to be very lightweight especially compared to before). 

4) The summary frequency does not seem to make a difference anymore (for SAC with self-play)

Any insight on any of these issues would be very appreciated. Thanks!

**Comment by niskander** [comment]
Might also be worth mentioning that even though only 1 team appears in the logs, both teams are receiving decisions correctly.

**Comment by alphonsocrawford** [comment]
@niskander 
For the first issue, seeing only Team 2 logs is actually intended behavior. When training with Self-Play, the first team is actually a fixed model, and second team is logged since that is the model that is being trained

For issue 2 and 4, we will look into these issues and follow-up as they could be bugs

And for issue 3, it is hard for us to reproduce an issue for an environment we did not create. Could you possibly reproduce this issue with one of our sample environments like Tennis?

**Comment by niskander** [comment]
@alphonsocrawford Thanks for the response.
For issue 1, thanks for the explanation. I looked through the code and understand how it's intended to work better now!

If the rewards from one team are ignored for the "Policy Extrinsic Reward" graph but both teams are considered for "Environment Cumulative Reward" it would explain the second issue (which wouldn't be an issue).

For the freeze, it's very random. I will try to see if I can repro it with one of the example environments. 

Still unsure about  the summary frequency, but it's not a big deal anyways. Thanks!

**Comment by niskander** [comment]
@alphonsocrawford I was investigating the freezing issues and noticed that the games with self-play take 4-5 times the normal simulation time! I turned it off for a quick sanity check (all else being the same) and the simulation time returned to normal. 

Have you experienced a similar issue in your tests? Any clue why it would take that much longer? (I'm referring to the run time of a single episode, not to training in general)

**Comment by niskander** [comment]
@alphonsocrawford Hi, I've done more investigating regarding these issues.

For the Policy Extrinsic Reward/Environment Cumulative Reward no longer being the same, I've encountered the issue even with self-play turned off. (I only have extrinsic rewards.) This behavior is definitely different from before and I believe it is a bug.

For the freezes, I believe these were caused by very frequent training updates using large batch sizes. I increased the batch sizes according to the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/0.14.0/docs/Migrating.md), which said that steps are now counted per agent (0.13 to 0.14)
I'm still unsure about the right number to use btw. To get one episode, I'm using `num_agents * num_envs * episode_length`
In any case, not related to Self-play either.

Should I be creating a different bug for the reward graphs issue?
